src/transport/tls/accept.rs

```diff
@@ -0,0 +1,280 @@
+use crate::core::listen::Accept;
+use crate::transport::prefixed::Prefixed;
+use crate::transport::tls::{self, conditional_accept, Connection, ReasonForNoPeerName};
+use crate::transport::{BoxedIo, GetOriginalDst};
+use crate::{identity, Conditional, Error};
+use bytes::BytesMut;
+use futures::{try_ready, Future, Poll};
+use indexmap::IndexSet;
+use linkerd2_dns_name as dns;
+pub use rustls::ServerConfig as Config;
+use std::net::SocketAddr;
+use std::sync::Arc;
+use tokio::{io::AsyncRead, net::TcpStream};
+use tracing::{debug, trace};
+
+pub trait HasConfig {
+    fn tls_server_name(&self) -> identity::Name;
+    fn tls_server_config(&self) -> Arc<Config>;
+}
+
+/// Produces a server config that fails to handshake all connections.
+pub fn empty_config() -> Arc<Config> {
+    let verifier = rustls::NoClientAuth::new();
+    Arc::new(Config::new(verifier))
+}
+
+pub struct AcceptTls<A: Accept<Connection>, T, G> {
+    accept: A,
+    tls: tls::Conditional<T>,
+    disable_protocol_detection_ports: IndexSet<u16>,
+    get_original_dst: G,
+}
+
+pub enum AcceptFuture<A: Accept<Connection>> {
+    TryTLS(Option<TryTLS<A>>),
```

Comment:
nit: I think 3 letter naming in Rust tends to be first letter uppercase--rest lowercase.